setup.py

```diff
@@ -0,0 +1,208 @@
+                            return f.read() == license_ref
+
+                    identical_flag = all(map(check_file, license_files[1:]))
+                    if not identical_flag:
+                        raise RuntimeError(
+                            "inconsistent license found for project: ",
+                            project_name,
+                            " check its license files under: ",
+                            license_files,
+                        )
+
+                    with open("LICENSE", "a") as f:
+                        f.write("\n\nProject Name: " + project_name)
+                        f.write("\nLicense Files:\n")
+                        for file_name in license_files:
+                            f.write("\t" + file_name)
+                        f.write("\n" + license_ref)
+
+                # generate whl before we restore LICENSE
+                super().run()
+
+def cmake():
+    # make build directories
+    build_dir_name = "build"
+    cwd = os.path.dirname(os.path.abspath(__file__))
+    cmake_build_dir = os.path.join(cwd, "build")
+    if not os.path.exists(cmake_build_dir):
+        os.makedirs(cmake_build_dir)
+
+    from tools.gen_nvfuser_version import get_pytorch_cmake_prefix
+    pytorch_cmake_config = "-DCMAKE_PREFIX_PATH=" + get_pytorch_cmake_prefix()
+
+    # generate cmake directory
+    cmd_str = [get_cmake_bin(), pytorch_cmake_config, "-B", build_dir_name, "."]
+    subprocess.check_call(cmd_str)
+
+    if not CMAKE_ONLY:
+        # build binary
+        max_jobs = str(multiprocessing.cpu_count())
+        cmd_str = [
+            get_cmake_bin(),
+            "--build",
+            build_dir_name,
+            "--target",
+            "install",
+            "--",
+            "-j",
+            max_jobs,
+        ]
+        subprocess.check_call(cmd_str)
+
+        # copy nvfuser pybind extension
+        src = os.path.join(cwd, "nvfuser", "lib", "libnvfuser.so")
+        dst = os.path.join(cwd, "nvfuser", "_C.cpython-310-x86_64-linux-gnu.so")
+        copy_file(src, dst)
+
+
+def main():
+    # NOTE: package include files for cmake
+    nvfuser_package_data = [
+        "*.so",
+        "lib/*.so",
+    ]
+
+    if BUILD_SETUP:
+        cmake()
+
+    from tools.gen_nvfuser_version import get_version
+
+    setup(
+        name="nvfuser",
+        # query nvfuser version
+        version=get_version(),
+        description="A Fusion Code Generator for NVIDIA GPUs (commonly known as 'nvFuser')",
+        packages=["nvfuser", "nvfuser_python_utils"],
+        license_files=("LICENSE",),
+        cmdclass={
+            "bdist_wheel": build_whl,
+            "clean": clean,
+        },
+        package_data={
+            "nvfuser": nvfuser_package_data,
+        },
+        entry_points={
+            'console_scripts': [
+                'patch-nvfuser = nvfuser_python_utils:patch_pytorch_nvfuser_binaries',
+            ],
+        },
```

Comment:
it might be better to explicitly mention pybind11[global] as a dependency
```suggestion
        },
        install_requires=["pybind11[global]"],
```

QQ: why is `[global]` needed? 

ref: https://setuptools.pypa.io/en/latest/userguide/dependency_management.html

Reply:
I believe `pybind11[global]` gives us both `pybind11` and `pybind11-global` with pip.

`pybind11-global` is needed for cmake, since it contains `Pybinbd11Confit.cmake`. pytorch doesn't ship pybind11 with its wheel.

Reply:
Can't really use setuptool for the requirement as suggested here. Since we need pybind11 for cmake, which happens before we call `setup`.
I added `requirements.txt`. so we should just run `pip install -r requirements.txt`.